CLI 编译命令生成器，在Ubuntu-20.04.1-x64系统，编译php8.2静态文件添加pgsql编译报错bug

背景：通过简单几步构建https://www.workerman.net/a/1605编译php8.2静态文件

起初是想构建的扩展如下：
![image](https://github.com/crazywhalecc/static-php-cli/assets/94685222/26d883d8-a24c-4999-ba87-ad822d312808)
重新编译这么多扩展时间较长，顾省略这里的报错

然后编译失败，尝试只编译一个扩展pgsql的报错如下
如下：
【编译到最后一步时候：】
使用CLI 编译命令生成器（https://static-php.dev/zh/guide/cli-generator.html）
在Ubuntu-20.04.1-x64系统
编译php8.2静态文件添加pgsql编译报错bug
报错如下：
CMake Error at CMakeLists.txt:1 (cmake_minimum_required):
  CMake 3.18 or higher is required.  You are running version 3.16.3

-- Configuring incomplete, errors occurred!
[22:47:56] [ERRO] Uncaught SPC\exception\RuntimeException: Command run failed with code[1]: cd '/home/root/custom_complie_php/source/libxml2/build' && cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/home/root/custom_complie_php/buildroot -DCMAKE_INSTALL_LIBDIR=/home/root/custom_complie_php/buildroot/lib -DCMAKE_TOOLCHAIN_FILE=/home/root/custom_complie_php/source/toolchain.cmake -DBUILD_SHARED_LIBS=OFF -DIconv_IS_BUILT_IN=OFF -DLIBXML2_WITH_ICONV=ON -DLIBXML2_WITH_ZLIB=ON -DZLIB_LIBRARY=/home/root/custom_complie_php/buildroot/lib/libz.a -DZLIB_INCLUDE_DIR=/home/root/custom_complie_php/buildroot/include -DLIBXML2_WITH_ICU=OFF -DLIBXML2_WITH_LZMA=OFF -DLIBXML2_WITH_PYTHON=OFF -DLIBXML2_WITH_PROGRAMS=OFF -DLIBXML2_WITH_TESTS=OFF .. at phar:///home/root/custom_complie_php/spc/src/globals/functions.php(100)
[22:47:56] [ERRO] #0 phar:///home/root/custom_complie_php/spc/src/SPC/util/UnixShell.php(47): f_passthru('cd '/home/root/...')
#1 phar:///home/root/custom_complie_php/spc/src/SPC/builder/linux/library/libxml2.php(27): SPC\util\UnixShell->exec('cd '/home/root/...')
#2 phar:///home/root/custom_complie_php/spc/src/SPC/builder/LibraryBase.php(144): SPC\builder\linux\library\libxml2->build()
#3 phar:///home/root/custom_complie_php/spc/src/SPC/builder/LibraryBase.php(152): SPC\builder\LibraryBase->tryBuild(true)
#4 phar:///home/root/custom_complie_php/spc/src/SPC/builder/unix/UnixBuilderBase.php(143): SPC\builder\LibraryBase->tryBuild(false)
#5 phar:///home/root/custom_complie_php/spc/src/SPC/command/BuildCliCommand.php(143): SPC\builder\unix\UnixBuilderBase->buildLibs(Array)
#6 phar:///home/root/custom_complie_php/spc/src/SPC/command/BaseCommand.php(99): SPC\command\BuildCliCommand->handle()
#7 phar:///home/root/custom_complie_php/spc/vendor/symfony/console/Command/Command.php(326): SPC\command\BaseCommand->execute(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#8 phar:///home/root/custom_complie_php/spc/vendor/symfony/console/Application.php(1078): Symfony\Component\Console\Command\Command->run(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#9 phar:///home/root/custom_complie_php/spc/vendor/symfony/console/Application.php(324): Symfony\Component\Console\Application->doRunCommand(Object(SPC\command\BuildCliCommand), Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#10 phar:///home/root/custom_complie_php/spc/vendor/symfony/console/Application.php(175): Symfony\Component\Console\Application->doRun(Object(Symfony\Component\Console\Input\ArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
#11 phar:///home/root/custom_complie_php/spc/bin/spc(20): Symfony\Component\Console\Application->run()
#12 /home/root/custom_complie_php/spc(8): require('phar:///home/ro...')
#13 {main}

你试一下 `cmake --version`，看看是不是 CMake 版本过旧了，因为我看你上面的报错信息中包含了：

```
CMake Error at CMakeLists.txt:1 (cmake_minimum_required):
CMake 3.18 or higher is required. You are running version 3.16.3
```

我用bin/spc-alpine-docker build --build-cli "apcu,bcmath,calendar,ctype,curl,dba,dom,event,exif,fileinfo,filter,gd,iconv,intl,mbregex,mbstring,mysqli,mysqlnd,opcache,openssl,pcntl,pdo,pdo_mysql,pdo_pgsql,pdo_sqlite,pgsql,phar,posix,readline,redis,session,simplexml,sockets,sodium,sqlite3,tokenizer,xml,xmlreader,xmlwriter,xsl,zip,zlib" --debug -I "memory_limit=512M"编译依旧报错如下：


[01:32:56] [DEBU] Running command with direct output: cd '/app/source/libxml2/build' &amp;&amp; cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/app/buildroot -DCMAKE_INSTALL_LIBDIR=/app/buildroot/lib -DCMAKE_TOOLCHAIN_FILE=/app/source/toolchain.cmake -DBUILD_SHARED_LIBS=OFF -DIconv_IS_BUILT_IN=OFF -DLIBXML2_WITH_ICONV=ON -DLIBXML2_WITH_ZLIB=ON -DZLIB_LIBRARY=/app/buildroot/lib/libz.a -DZLIB_INCLUDE_DIR=/app/buildroot/include -DLIBXML2_WITH_ICU=ON -DLIBXML2_WITH_LZMA=OFF -DLIBXML2_WITH_PYTHON=OFF -DLIBXML2_WITH_PROGRAMS=OFF -DLIBXML2_WITH_TESTS=OFF ..
Error relocating /usr/bin/cmake: SSL_get0_group_name: symbol not found
[01:32:56] [ERRO] Uncaught SPC\exception\RuntimeException: Command run failed with code[127]: cd '/app/source/libxml2/build' &amp;&amp; cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/app/buildroot -DCMAKE_INSTALL_LIBDIR=/app/buildroot/lib -DCMAKE_TOOLCHAIN_FILE=/app/source/toolchain.cmake -DBUILD_SHARED_LIBS=OFF -DIconv_IS_BUILT_IN=OFF -DLIBXML2_WITH_ICONV=ON -DLIBXML2_WITH_ZLIB=ON -DZLIB_LIBRARY=/app/buildroot/lib/libz.a -DZLIB_INCLUDE_DIR=/app/buildroot/include -DLIBXML2_WITH_ICU=ON -DLIBXML2_WITH_LZMA=OFF -DLIBXML2_WITH_PYTHON=OFF -DLIBXML2_WITH_PROGRAMS=OFF -DLIBXML2_WITH_TESTS=OFF .. at /app/src/globals/functions.php(100)
[01:32:56] [ERRO] #0 /app/src/SPC/util/UnixShell.php(47): f_passthru()
#1 /app/src/SPC/builder/linux/library/libxml2.php(27): SPC\util\UnixShell-&gt;exec()
#2 /app/src/SPC/builder/LibraryBase.php(144): SPC\builder\linux\library\libxml2-&gt;build()
#3 /app/src/SPC/builder/LibraryBase.php(152): SPC\builder\LibraryBase-&gt;tryBuild()
#4 /app/src/SPC/builder/unix/UnixBuilderBase.php(143): SPC\builder\LibraryBase-&gt;tryBuild()
#5 /app/src/SPC/command/BuildCliCommand.php(143): SPC\builder\unix\UnixBuilderBase-&gt;buildLibs()
#6 /app/src/SPC/command/BaseCommand.php(99): SPC\command\BuildCliCommand-&gt;handle()
#7 /app/vendor/symfony/console/Command/Command.php(326): SPC\command\BaseCommand-&gt;execute()
#8 /app/vendor/symfony/console/Application.php(1078): Symfony\Component\Console\Command\Command-&gt;run()
#9 /app/vendor/symfony/console/Application.php(324): Symfony\Component\Console\Application-&gt;doRunCommand()
#10 /app/vendor/symfony/console/Application.php(175): Symfony\Component\Console\Application-&gt;doRun()
#11 /app/bin/spc(21): Symfony\Component\Console\Application-&gt;run()
#12 {main}





版本如下：
***@***.***:/home/root/git/static-php-cli# cmake --version
cmake version 3.10.2







＆肀爱★
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "crazywhalecc/static-php-cli"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年3月25日(星期一) 下午4:05
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [crazywhalecc/static-php-cli] CLI 编译命令生成器，在Ubuntu-20.04.1-x64系统，编译php8.2静态文件添加pgsql编译报错bug (Issue #396)





 
你试一下 cmake --version，看看是不是 CMake 版本过旧了，因为我看你上面的报错信息中包含了：
 CMake Error at CMakeLists.txt:1 (cmake_minimum_required): CMake 3.18 or higher is required. You are running version 3.16.3  
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you authored the thread.Message ID: ***@***.***&gt;

--- a/src/SPC/doctor/item/LinuxToolCheckList.php
+++ b/src/SPC/doctor/item/LinuxToolCheckList.php
@@ -74,6 +74,22 @@ class LinuxToolCheckList
         return CheckResult::ok();
     }
 
+    #[AsCheckItem('if cmake version >= 3.18', limit_os: 'Linux')]
+    public function checkCMakeVersion(): ?CheckResult
+    {
+        $check_cmd = 'cmake --version';
+        $pattern = '/cmake version (.*)/m';
+        $out = shell()->execWithResult($check_cmd, false)[1][0];
+        if (preg_match($pattern, $out, $match)) {
+            $ver = $match[1];
+            if (version_compare($ver, '3.18.0') <= 0) {
+                return CheckResult::fail('cmake version is too low (' . $ver . '), please update it manually!');
+            }
+            return CheckResult::ok($match[1]);
+        }
+        return CheckResult::fail('Failed to get cmake version');
+    }
+
     /** @noinspection PhpUnused */
     #[AsCheckItem('if necessary linux headers are installed', limit_os: 'Linux')]
     public function checkSystemOSPackages(): ?CheckResult

--- a/src/SPC/doctor/item/OSCheckList.php
+++ b/src/SPC/doctor/item/OSCheckList.php
@@ -20,7 +20,7 @@ class OSCheckList
             return CheckResult::fail('Current OS is not supported: ' . PHP_OS_FAMILY);
         }
         $distro = PHP_OS_FAMILY === 'Linux' ? (' ' . SystemUtil::getOSRelease()['dist']) : '';
-        $known_distro = PHP_OS_FAMILY === 'Linux' && in_array(SystemUtil::getOSRelease()['dist'], SystemUtil::getSupportedDistros());
+        $known_distro = PHP_OS_FAMILY !== 'Linux' || in_array(SystemUtil::getOSRelease()['dist'], SystemUtil::getSupportedDistros());
         return CheckResult::ok(PHP_OS_FAMILY . ' ' . php_uname('m') . $distro . ', supported' . ($known_distro ? '' : ' (but not tested on this distro)'));
     }
 }

